source/MaterialXTest/ShaderGen.cpp

```diff
@@ -868,3 +961,239 @@ TEST_CASE("LayeredSurface", "[shadergen]")
+        const std::string target("");
+
+        std::string nodeDefName = nodeDef->getName();
+        std::string nodeName = nodeDef->getNodeString();
+        if (!requiresImplementation(nodeDef))
+        {
+            found_str += "No implementation required for nodedef: " + nodeDefName + ", Node: "
+                + nodeName + ". Namespace: " + nodeDef->getAttribute("namespace") + "\n";
+            continue;
+        }
+
+        mx::ImplementationPtr impl =
+            findImplementation(doc, nodeDefName, language, target);
+        if (!impl)
+        {
+            missing++;
+            missing_str += "Missing nodeDef implemenation: " + nodeDefName + ", Node: " + nodeName
+                + ". Namespace: " + nodeDef->getAttribute("namespace") + "\n";
+        }
+        else
+        {
+            // Scan for file and see if we can read in the contents
+            std::string sourceContents;
+            MaterialX::FilePath sourcePath = impl->getFile();
+            MaterialX::FilePath resolvedPath = sourceCodeSearchPath.find(sourcePath);
+            bool found = MaterialX::readFile(resolvedPath.asString(), sourceContents);
+            if (!found)
+            {
+                missing++;
+                missing_str += "Missing source code: " + sourcePath.asString() + " for nodeDef: "
+                    + nodeDefName + ". Impl: " + impl->getName() + ". Namespace: " + nodeDef->getAttribute("namespace") + "\n";
+            }
+            else
+            {
+                found_str += "Found impl and src for nodedef: " + nodeDefName + ", Node: "
+                    + nodeName + ". Impl: " + impl->getName() + ". Namespace: " + nodeDef->getAttribute("namespace") + "\n";
+            }
+        }
+    }
+
+    implDumpStream << "-----------------------------------------------------------------------" << std::endl;
+    implDumpStream << "Missing: " << missing << " implementations out of: " << count << " nodedefs\n";
+    implDumpStream << missing_str << std::endl;
+    implDumpStream << found_str << std::endl;
+    implDumpStream << "-----------------------------------------------------------------------" << std::endl;
+
+    implDumpBuffer.close();
+}
+
+TEST_CASE("Shadergen implementation validity", "[shadergen]")
+{
+    mx::ObservedDocumentPtr doc = mx::Document::createDocument<mx::ObservedDocument>();
+
+    // Load standard libraries
+    std::vector<std::string> filenames =
+    {
+        "documents/Libraries/stdlib/mx_stdlib_defs.mtlx",
+        "documents/Libraries/sx/sx_defs.mtlx",
+        "documents/Libraries/adsk/adsk_defs.mtlx"
+    };
+    std::vector<std::string> libraryNames =
+    {
+        "stdlib",
+        "sx",
+        "adsk"
+    };
+    std::shared_ptr<LibraryObserver> observer = std::make_shared<LibraryObserver>();
+    doc->addObserver("libraryObserver", observer);
+    for (size_t i=0; i<filenames.size(); i++)
+    {
+        observer->setLibraryRefName(libraryNames[i]);
```

Comment:
This is the temporary tagging I put in to add the library name as namespace for these tests. This done explicitly in our utility library manager.